### PR TITLE
Fix ICE when trying to decode too large static arrays

### LINKED ISSUE #8923: [Typechecker] ICE because array size cannot fit unsigned type
## Description

```
contract C {
        function f() public {
                abi.decode("", (byte[999999999]));
        }
}
```

throws

https://github.com/ethereum/solidity/blob/b2122d479fc0554a9506f4e393fe95aa52f503d3/libsolidity/ast/Types.cpp#L1727

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ solc <input.sol>
```

May be related to #8665

### PR DESCRIPTION
**Pull request by Marenz**
fixes #8923

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -193,6 +193,18 @@ TypePointers TypeChecker::typeCheckABIDecodeAndRetrieveReturnType(FunctionCall c
 					typeArgument->location(),
 					"Decoding type " + actualType->toString(false) + " not supported."
 				);
+
+			if (auto referenceType = dynamic_cast<ArrayType const*>(actualType))
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Why only for array type and not for all reference types?

Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp]: Good point. Will change.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -193,6 +193,18 @@ TypePointers TypeChecker::typeCheckABIDecodeAndRetrieveReturnType(FunctionCall c
 					typeArgument->location(),
 					"Decoding type " + actualType->toString(false) + " not supported."
 				);
+
+			if (auto referenceType = dynamic_cast<ReferenceType const*>(actualType))
+			{
+				auto result = referenceType->validForLocation(referenceType->location());
```
Comment (leonardoalt) [inline libsolidity/analysis/TypeChecker.cpp:199]: Can be inlined inside the `if`

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:199]: Don't see how - it is negated.

**Review by chriseth — APPROVED** [review]

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/analysis/TypeChecker.cpp:199]: What's wrong with
`if (!referenceType->validForLocation(referenceType->location()))`
?

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:199]: `result` is needed further down

Comment (leonardoalt) [reply libsolidity/analysis/TypeChecker.cpp:199]: missed that :p

**Pull request merged**